File: surf.cpp
#include <opencv2/xfeatures2d.hpp>
#include <opencv2/features2d.hpp>
#include <opencv2/core.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/highgui.hpp>

#include <iostream>
#include <memory>
#include <fstream>

int main() {
    std::vector<cv::KeyPoint> keypoints;
    cv::Ptr<cv::Feature2D> surf = cv::xfeatures2d::SURF::create();
    cv::Mat image;
    
    image = cv::imread("./lena.png");
    
    cv::Mat mask = cv::Mat(image.size().height, 
                           image.size().width,
                           CV_8UC1, cv::Scalar(255));
    
    cv::Mat descriptors;
    
    surf->detectAndCompute(image, mask, keypoints, descriptors);
    
    cv::drawKeypoints(image, keypoints, image);
    cv::imshow("image", image);
    
    // save to a csv all the descriptors
    std::ofstream out("surf_descriptors.csv");
    for(int i=0; i<descriptors.rows; i++){
        for(int j=0; j<descriptors.cols; j++){
                out << (float)descriptors.at<float>(i, j) <<',';
        }
        out << '\n';
    }
    
    std::cout<<"There are "<<descriptors.rows<<" keypoints"<<std::endl;
    std::cout<<"Each surf descriptor has "<<descriptors.cols<<" values"<<std::endl;
  
    cv::waitKey(0);
    return 0;
}

File: surf.py
import cv2
import numpy as np
import argparse

def main(image_path):
    
    img = cv2.imread(image_path)

    # Create SURF object with Hessian Threshold = 400
    surf = cv2.xfeatures2d.SURF_create(400)

    # compute the keypoint and feature descriptors
    kp, des = surf.detectAndCompute(img, None)

    # display the results
    img2 = cv2.drawKeypoints(img, kp, None, (255,0,0), 4)
    cv2.imshow("image", img2)
    
    # save the descriptors in a csv file
    np.savetxt("surf_descriptors.csv", des, delimiter=",")

    # close when a key is pressed
    cv2.waitKey(0)


# running example:
# python surf.py --image lena.png
if __name__ == "__main__":
    
    parser = argparse.ArgumentParser(
            description="Save to csv the surf descriptors of the given image"
            )
    
    parser.add_argument("--image", required=True, help="path to the image")
    args = parser.parse_args()
    
    main(image_path=args.image)
